client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java

```diff
@@ -177,22 +128,26 @@ public void close() {
     }
 
     @Override
-    public ReadFuture read(long offset, int length) {
+    public CompletableFuture<SegmentRead> read(long offset, int length) {
         Exceptions.checkNotClosed(closed.get(), this);
         WireCommands.ReadSegment request = new WireCommands.ReadSegment(segmentId.getScopedName(), offset, length);
-        
-        ReadFutureImpl read = new ReadFutureImpl(request);
+
+        return backoffSchedule.retryingOn(ConnectionFailedException.class)
+                              .throwingOn(ConnectionClosedException.class)
+                              .runAsync(() -> {
+                                  return getConnection().thenCompose(c -> sendRequestOverConnection(request, c));
```

Comment:
If `getConnection() `fails then we need to convert the exception to `ConnectionFailedException`, so that retry kicks in.